models/userStatus.js

```diff
@@ -78,39 +79,136 @@ const getAllUserStatus = async (query) => {
+          const tommorow = getTommorowTimeStamp();
+          if (newStatusData.currentStatus.from >= tommorow) {
+            newStatusData.futureStatus = { ...newStatusData.currentStatus };
+            delete newStatusData.currentStatus;
+          } else {
+            newStatusData.futureStatus = {};
+          }
         }
       }
-      await userStatusModel.doc(docId).update(updatedData);
-      return { id: docId, userStatusExists: true, data: updatedData };
+      await userStatusModel.doc(docId).update(newStatusData);
+      return { id: docId, userStatusExists: true, data: newStatusData };
     } else {
       // the user doc doesnt exist meaning we need to create one
-      const { id } = await userStatusModel.add({ userId, ...updatedData });
-      return { id, userStatusExists: false, data: updatedData };
+      if (Object.keys(newStatusData).includes("currentStatus")) {
+        const newUserState = newStatusData.currentStatus.state;
+        const isNewStateOOO = newUserState === userState.OOO;
+        const isNewStateActive = newUserState === userState.ACTIVE;
+        if (!isNewStateOOO) {
+          newStatusData.currentStatus.until = "";
+        }
+        if (isNewStateActive) {
+          newStatusData.currentStatus.message = "";
+        }
+        if (isNewStateOOO) {
+          const tommorow = getTommorowTimeStamp();
+          if (newStatusData.currentStatus.from >= tommorow) {
+            const futureStatus = { ...newStatusData.currentStatus };
+            delete newStatusData.currentStatus;
+            newStatusData.futureStatus = futureStatus;
+          }
+        }
+      }
+      const { id } = await userStatusModel.add({ userId, ...newStatusData });
+      return { id, userStatusExists: false, data: newStatusData };
     }
   } catch (error) {
     logger.error(`error in updating User Status Document ${error}`);
     throw error;
   }
 };
 
-module.exports = { deleteUserStatus, getUserStatus, getAllUserStatus, updateUserStatus };
+/**
+ * @param userId { String }: Id of the User
+ * @param newStatusData { Object }: Data to be Updated
+ * @returns Promise<userStatusModel|Object>
+ */
+
+const updateAllUserStatus = async () => {
+  try {
+    const userStatusDocs = await userStatusModel.where("futureStatus.state", "in", ["ACTIVE", "IDLE", "OOO"]).get();
+    const batch = firestore.batch();
+    const today = new Date().getTime();
+    userStatusDocs.forEach(async (document) => {
+      const doc = document.data();
+      const docRef = document.ref;
+      const newStatusData = { ...doc };
+      let toUpdate = false;
+      const { futureStatus, currentStatus } = doc;
+      const { state: futureState } = futureStatus;
+      if (futureState === "ACTIVE" || futureState === "IDLE") {
+        if (today >= futureStatus.from) {
+          newStatusData.currentStatus = { ...futureStatus, until: "", updatedAt: today };
+          newStatusData.futureStatus = {};
+          toUpdate = !toUpdate;
```

Comment:
because this is inside the`if block` it makes more sense to assign a boolean to `toUpdate` instead of toggling it. Applies to other places as well